wolfcrypt/src/port/Espressif/esp32_mp.c

```diff
@@ -0,0 +1,488 @@
+    DPORT_REG_WRITE(RSA_INTERRUPT_REG, 1);
+    /* start process  */
+    DPORT_REG_WRITE(reg, 1);
+}
+/* wait until done */
+static int wait_uitil_done(word32 reg)
+{
+    int timeout = 0;
+    /* wait until done && not timeout */
+    while(1) {
+        if(++timeout < ESP_RSA_TIMEOUT && DPORT_REG_READ(reg) == 1){
+            break;
+        }
+    }
+
+    /* clear interrupt */
+    DPORT_REG_WRITE(RSA_INTERRUPT_REG, 1);
+
+    if(timeout >= ESP_RSA_TIMEOUT) {
+        ESP_LOGE(TAG, "rsa operation is time-outed.");
+        return MP_NG;
+    }
+
+    return MP_OKAY;
+}
+/* read data from memory into mp_init          */
+static void esp_memblock_to_mpint(word32 mem_address, mp_int* mp, word32 numwords)
+{
+    esp_dport_access_read_buffer((uint32_t*)mp->dp, mem_address, numwords);
+    mp->used = numwords;
+}
+
+/* write mp_init into memory block             */
+static void esp_mpint_to_memblock(word32 mem_address, const mp_int* mp,
+                                                      const word32 bits,
+                                                      const word32 hwords)
+{
+    word32 i;
+    word32 len = (bits / 8 + ((bits & 7) != 0 ? 1 : 0));
+
+    len = (len+sizeof(word32)-1)/sizeof(word32);
+
+    for(i=0;i < hwords; i++) {
+        if(i < len) {
+            DPORT_REG_WRITE(mem_address + (i * sizeof(word32)), mp->dp[i]);
+        } else {
+            DPORT_REG_WRITE(mem_address + (i * sizeof(word32)), 0);
+        }
+    }
+}
+/* return needed hw words.                                 */
+/* supported words length                                  */
+/*  words : {16 ,  32,  48,    64,   80,   96, 112,   128} */
+/*  bits  : {512,1024, 1536, 2048, 2560, 3072, 3584, 4096} */
+static word32 words2hwords(word32 wd)
+{
+    const word32 shit_ = 4;
+
+    return (((wd + 0xf)>>shit_)<<shit_);
+}
+/* count the number of words is needed for bits */
+static word32 bits2words(word32 bits)
+{
+    /* 32 bits */
+    const word32 d = sizeof(word32) * WOLFSSL_BIT_SIZE;
+
+    return((bits + (d - 1))/d);
+}
+/* get rinv */
+static int esp_get_rinv(mp_int *rinv, mp_int *M, word32 exp)
+{
+    int ret = 0;
+
+    /* 2^(exp)*/
+    if((ret = mp_2expt(rinv, exp)) != MP_OKAY) {
+        ESP_LOGE(TAG, "failed to calculate mp_2expt()");
+        return ret;
+    }
+
+    /* r_inv = R^2 mod M(=P) */
+    if(ret == 0 && (ret = mp_mod(rinv, M, rinv)) != MP_OKAY){
+        ESP_LOGE(TAG, "failed to calculate mp_mod()");
+        return ret;
+    }
+
+    return ret;
+}
+/* Z = X * Y;                                              */
+int esp_mp_mul(fp_int* X, fp_int* Y, fp_int* Z)
+{
+    int ret = 0;
+    int neg = (X->sign == Y->sign)? MP_ZPOS : MP_NEG;
+
+    word32 Xs;
+    word32 Ys;
+    word32 Zs;
+    word32 maxWords_sz;
+    word32 hwWords_sz;
+
+    /* ask bits number */
+    Xs = mp_count_bits(X);
+    Ys = mp_count_bits(Y);
+    Zs = Xs + Ys;
+
+    /* maximum bits and words for writing to hw */
+    maxWords_sz = bits2words(MAX(Xs, Ys));
```

Comment:
Consider using our `max()` and misc.h as indicated previously.

Reply:
Changed to use max() and min() in misc.h
Thank you for pointing out.